File: backend/internal/types/jobs.go
package types

type Job struct {
	JobID         int    `json:"job_id"`
	Name          string `json:"name"`
	UserName      string `json:"user_name"`
	Partition     string `json:"partition"`
	State         string `json:"state"`
	SubmitTime    int64  `json:"submit_time"`
	StartTime     int64  `json:"start_time"`
	EndTime       int64  `json:"end_time"`
	Nodes         string `json:"nodes"`
	CPUs          int    `json:"cpus"`
	MemoryPerNode int    `json:"memory_per_node"`
}

type JobResponse struct {
	Jobs []Job `json:"jobs"`
}

type JobResponseSingle struct {
	Job Job `json:"job"`
}

type TimeInfo struct {
	Elapsed    int64 `json:"elapsed"`
	Eligible   int64 `json:"eligible"`
	End        int64 `json:"end"`
	Start      int64 `json:"start"`
	Submission int64 `json:"submission"`
}

type JobResource struct {
	Type  string `json:"type"`
	Name  string `json:"name"`
	ID    int    `json:"id"`
	Count int    `json:"count"`
}

type JobComment struct {
	Administrator string `json:"administrator"`
	Job           string `json:"job"`
	System        string `json:"system"`
}
type JobState struct {
	Current []string `json:"current"`
	Reason  string   `json:"reason"`
}

type TresResource struct {
	Type  string `json:"type"`
	Name  string `json:"name"`
	ID    int    `json:"id"`
	Count int    `json:"count"`
}

type JobDetails struct {
	Account          string     `json:"account"`
	Comment          JobComment `json:"comment"`
	AllocationNodes  int        `json:"allocation_nodes"`
	Cluster          string     `json:"cluster"`
	JobID            int        `json:"job_id"`
	Name             string     `json:"name"`
	Nodes            string     `json:"nodes"`
	Partition        string     `json:"partition"`
	QOS              string     `json:"qos"`
	State            JobState   `json:"state"`
	SubmitLine       string     `json:"submit_line"`
	User             string     `json:"user"`
	WorkingDirectory string     `json:"working_directory"`
	Time             struct {
		Elapsed    int64 `json:"elapsed"`
		Eligible   int64 `json:"eligible"`
		End        int64 `json:"end"`
		Start      int64 `json:"start"`
		Submission int64 `json:"submission"`
	} `json:"time"`
	Tres struct {
		Allocated []TresResource `json:"allocated"`
		Requested []TresResource `json:"requested"`
	} `json:"tres"`
}

type JobDetailsResponse struct {
	Jobs []JobDetails `json:"jobs"`
}

File: src/shared/api/v1/jobs.ts
// Public v1 jobs contract: pure types, camelCase, explicit units, ISO timestamps.
// JOB_STATES is the closed normalized state vocabulary shared by server and client.
export const JOB_STATES = [
  'BOOT_FAIL',
  'CANCELLED',
  'COMPLETED',
  'DEADLINE',
  'FAILED',
  'NODE_FAIL',
  'OUT_OF_MEMORY',
  'PENDING',
  'PREEMPTED',
  'RUNNING',
  'SUSPENDED',
  'TIMEOUT',
  'UNKNOWN',
] as const;

export type JobState = (typeof JOB_STATES)[number];

export type TimeLimitDto = { kind: 'finite'; seconds: number } | { kind: 'infinite' } | null;

export interface GpuBreakdownDto {
  total: number;
  byType: Record<string, number>;
}

export interface RequestedResourcesDto {
  cpus: number | null;
  memoryMiB: number | null;
  nodes: number | null;
  gpus: GpuBreakdownDto;
}

export interface AllocatedResourcesDto extends RequestedResourcesDto {
  nodes: number | null;
}

export interface JobDto {
  id: string;
  jobId: string;
  arrayJobId: string | null;
  arrayTaskId: string | null;
  partition: string | null;
  name: string | null;
  user: string | null;
  account: string | null;
  qos: string | null;
  state: JobState;
  stateFlags: string[];
  stateReason: string | null;
  timeLimit: TimeLimitDto;
  submitTime: string | null;
  eligibleTime: string | null;
  startTime: string | null;
  endTime: string | null;
  priority: number | null;
  taskCount: number | null;
  cpusPerTask: number | null;
  constraints: string | null;
  reservation: string | null;
  nodeCount: number | null;
  nodeExpression: string | null;
  requested: RequestedResourcesDto;
  allocated: AllocatedResourcesDto;
  workdir: string | null;
  command: string | null;
  stdoutPath: string | null;
  stderrPath: string | null;
  dependency: string | null;
  exitCode: string | null;
  derivedExitCode: string | null;
  wckey: string | null;
  batchHost: string | null;
  flags: string[];
}

export interface JobsPaginationDto {
  page: number;
  pageSize: number;
  totalItems: number;
  totalPages: number;
}

export interface JobsResponse {
  jobs: JobDto[];
  pagination: JobsPaginationDto;
  updatedAt: string;
}

// Detail reuses JobDto so list and page share one normalization path.
export interface JobDetailsResponse {
  job: JobDto;
  updatedAt: string;
}

// Canonical IDs the jobs API can emit: a job id, or an array task id.
// Step suffixes (".batch", ".extern", ".0") are not part of this space.
export const CANONICAL_JOB_ID_PATTERN = /^[0-9]+(_[0-9]+)?$/;

export const JOBS_PAGE_DEFAULT = 1;
export const JOBS_PAGE_SIZE_DEFAULT = 20;
export const JOBS_PAGE_SIZE_MAX = 100;
